server/mediaserver.go

```diff
@@ -285,143 +259,13 @@ func gotRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
-							// XXX handle case when orch expects direct upload
-							uri, err := ios.SaveData(name, seg.Data)
-							if err != nil {
-								glog.Error("Error saving segment to OS ", err)
-								if monitor.Enabled {
-									monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
-								}
-								return
-							}
-							seg.Name = uri // hijack seg.Name to convey the uploaded URI
-						}
-
-						// send segment to the orchestrator
-						glog.V(common.DEBUG).Infof("Submitting segment %d", seg.SeqNo)
-
-						res, err := SubmitSegment(sess, seg, nonce)
-						if err != nil {
-							if shouldStopStream(err) {
-								glog.Warningf("Stopping current stream due to: %v", err)
-								rtmpStrm.Close()
-							}
-							return
-						}
-
-						// download transcoded segments from the transcoder
-						gotErr := false // only send one error msg per segment list
-						errFunc := func(subType, url string, err error) {
-							glog.Errorf("%v error with segment %v: %v (URL: %v)", subType, seg.SeqNo, err, url)
-							if monitor.Enabled && !gotErr {
-								monitor.LogSegmentTranscodeFailed(subType, nonce, seg.SeqNo, err)
-								gotErr = true
-							}
-						}
-
-						segHashes := make([][]byte, len(res.Segments))
-						n := len(res.Segments)
-						SegHashLock := &sync.Mutex{}
-						cond := sync.NewCond(SegHashLock)
-
-						dlFunc := func(url string, i int) {
-							defer func() {
-								cond.L.Lock()
-								n--
-								if n == 0 {
-									cond.Signal()
-								}
-								cond.L.Unlock()
-							}()
-
-							if bos := sess.BroadcasterOS; bos != nil && !drivers.IsOwnExternal(url) {
-								data, err := drivers.GetSegmentData(url)
-								if err != nil {
-									errFunc("Download", url, err)
-									return
-								}
-								name := fmt.Sprintf("%s/%d.ts", sess.Profiles[i].Name, seg.SeqNo)
-								newUrl, err := bos.SaveData(name, data)
-								if err != nil {
-									errFunc("SaveData", url, err)
-									return
-								}
-								url = newUrl
-
-								hash := crypto.Keccak256(data)
-								SegHashLock.Lock()
-								segHashes[i] = hash
-								SegHashLock.Unlock()
-							}
-
-							if monitor.Enabled {
-								monitor.LogTranscodedSegmentAppeared(nonce, seg.SeqNo, sess.Profiles[i].Name)
-							}
-							err = cpl.InsertHLSSegment(&sess.Profiles[i], seg.SeqNo, url, seg.Duration)
-							if err != nil {
-								errFunc("Playlist", url, err)
-								return
-							}
-						}
-
-						for i, v := range res.Segments {
-							go dlFunc(v.Url, i)
-						}
-
-						cond.L.Lock()
-						for n != 0 {
-							cond.Wait()
-						}
-						cond.L.Unlock()
-
-						// if !eth.VerifySig(transcoderAddress, crypto.Keccak256(segHashes...), res.Sig) { // need transcoder address here
-						// 	glog.Error("Sig check failed for segment ", seg.SeqNo)
-						// 	return
-						// }
-
-						glog.V(common.DEBUG).Info("Successfully validated segment ", seg.SeqNo)
-					}()
-				}
+				processSegment(cxn, seg)
```

Comment:
How do we know `cxn.sess` will be set here? I see that we're setting it on line 292, through `go s.startSessionListener(cxn)`, but will that for sure execute before this line?

Reply:
We don't; `processSegment` returns early if `cxn.sess` is nil. There are situations where this is normal, eg view-only streams or while we are in the middle of the O re-selection process for failover.  https://github.com/livepeer/go-livepeer/pull/646/files#diff-cecc27bfe2b6b19181ce443733dfba57R93

Where we set `sess = cxn.sess` in `processSegment` should probably be wrapped in a mutex though.

Reply:
That makes sense! So in which situations would the `sess` be set here? Doesn't `go s.startSessionListener(cxn)` only run once (and I think that's the only place where we're setting `sess`?? )? Wouldn't line 292 always run after `processSegment(cxn, seg)`?